### PR TITLE
Add assertion checking if too many args are passed into exports.

### LINKED ISSUE #21348: Add assertions for number of arguments for exported functions
Requested in another forum...
It'd be helpful to have debug assertions for how many arguments are passed into exports (EXPORTED_FUNCTIONS/EMSCRIPTEN_KEEPALIVE). Seems like it'd be relatively easy to modify `createExportWrapper`.

### PR DESCRIPTION
**Pull request by brendandahl**
Fixes #21348

### CODE REVIEW

**Comment by brendandahl** [comment]
I'm guessing this will pick up some failures in other places...

File: src/preamble.js [hunk 1]
```diff
@@ -512,14 +512,15 @@ Module['FS_createPreloadedFile'] = FS.createPreloadedFile;
...
-function createExportWrapper(name) {
+function createExportWrapper(name, nargs) {
   return (...args) => {
     assert(runtimeInitialized, `native function \`${name}\` called before runtime initialization`);
 #if EXIT_RUNTIME
     assert(!runtimeExited, `native function \`${name}\` called after runtime exit (use NO_EXIT_RUNTIME to keep it alive after main() exits)`);
 #endif
     var f = wasmExports[name];
     assert(f, `exported native function \`${name}\` not found`);
+    assert(args.length === nargs, `native function \`${name}\` called with ${args.length} args but expected ${nargs}`);
```
Comment (sbc100) [inline src/preamble.js:523]: s/expected/expects/?

**Review by sbc100 — COMMENTED** [review]
Nice!

I wonder if we will see any tests fail because of this :)

There could be some cases where a developer might be relying on the VM to automatically inject zero values for missing arguments.. so it might be safer to only report too many arguments and not too few?

File: test/test_other.py [hunk 2]
```diff
@@ -11985,6 +11985,20 @@ def test_native_call_after_exit(self):
...
     self.assertContained('native function `main` called after runtime exit', out)
 
+  def test_native_call_nargs(self):
+    self.set_setting('ASSERTIONS')
+    self.set_setting('EXPORTED_FUNCTIONS', ['_main', '_foo'])
+    create_file('foo.c', r'''
+      #include <emscripten.h>
+      void foo(int arg) {}
+      int main() {
+        EM_ASM(Module._foo(99, 100));
```
Comment (sbc100) [inline test/test_other.py:11995]: Just `_foo` works here I think,

**Comment by brendandahl** [comment]
> I wonder if we will see any tests fail because of this :)
> 
> There could be some cases where a developer might be relying on the VM to automatically inject zero values for missing arguments.. so it might be safer to only report too many arguments and not too few?

Looks like a number of tests fail where we rely on that behavior. I'll change it to just check for too many arguments.

**Comment by sbc100** [comment]
> > I wonder if we will see any tests fail because of this :)
> > There could be some cases where a developer might be relying on the VM to automatically inject zero values for missing arguments.. so it might be safer to only report too many arguments and not too few?
> 
> Looks like a number of tests fail where we rely on that behavior. I'll change it to just check for too many arguments.

I know ASYNCIFY relies on this.. was there anything else?

**Comment by brendandahl** [comment]
I've looked through a few and it seems the majority come from `_emscripten_thread_init`.

**Comment by brendandahl** [comment]
We could try to fix them to be "more correct", but code size could be worse and it would change the behavior of existings APIs. Any thoughts?

**Comment by sbc100** [comment]
> I've looked through a few and it seems the majority come from `_emscripten_thread_init`.

Is that because the entry point for a pthread sometimes has zero args and sometimes as 1 argument?  I think that is hard to work around sadly :(      Luckily the zero arg form is technically undefined behaviour so we could just disallow that.

**Comment by sbc100** [comment]
> We could try to fix them to be "more correct", but code size could be worse and it would change the behavior of existings APIs. Any thoughts?

I don't think its worth adding complexity but perhaps there are some cases that are actually bugs.

**Comment by brendandahl** [comment]
> > I've looked through a few and it seems the majority come from `_emscripten_thread_init`.
> 
> Is that because the entry point for a pthread sometimes has zero args and sometimes as 1 argument? I think that is hard to work around sadly :( Luckily the zero arg form is technically undefined behaviour so we could just disallow that.

It's mainly from here: https://github.com/emscripten-core/emscripten/blob/a95c44ee924d02fa1498f846595485d27c31daa8/src/worker.js#L232 we don't pass in the last two arguments `int default_stacksize, int start_profiling`

**Comment by sbc100** [comment]
> > > I've looked through a few and it seems the majority come from `_emscripten_thread_init`.
> > 
> > 
> > Is that because the entry point for a pthread sometimes has zero args and sometimes as 1 argument? I think that is hard to work around sadly :( Luckily the zero arg form is technically undefined behaviour so we could just disallow that.
> 
> It's mainly from here:
> 
> https://github.com/emscripten-core/emscripten/blob/a95c44ee924d02fa1498f846595485d27c31daa8/src/worker.js#L232
> 
> we don't pass in the last two arguments `int default_stacksize, int start_profiling`

That one might be easy enough to fix..

File: tools/webidl_binder.py [hunk 3]
```diff
@@ -546,7 +547,7 @@ def is_ptr_arg(i):
   c_names = {}
 
   def make_call_args(i):
-    if pre_arg:
+    if has_pre_arg:
```
Comment (sbc100) [inline tools/webidl_binder.py:550]: I don't see how this is any different to `bool(pre_arg)` which itself is identical to `pre_arg` in an `if` statement, no?

Comment (brendandahl) [reply tools/webidl_binder.py:550]: There's an assignment above that does `call_args = pre_arg`, then we append to `call_args` which modifies pre_arg as well.

Comment (brendandahl) [reply tools/webidl_binder.py:550]: Alternatively, we could instead clone/copy pre_arg to call_args so modifying one does not modify the other.

Comment (sbc100) [reply tools/webidl_binder.py:550]: Ah, in that case I think its better to do `call_args = list(pre_arg)` or `call_args = copy(pre_arg)` above.

Comment (sbc100) [reply tools/webidl_binder.py:550]: oops `call_args = pre_arg.copy()`

**Review by sbc100 — APPROVED** [review]
LGTM.  

Maybe update the PR title/descrption to mention that we only check for too-many and not too-few and why (because there can be valid use case for passing too few).  Maybe add a comment before the assertion to that effect too?

**Review by sbc100 — APPROVED** [review]

**Pull request merged**